--- a/lib/libxdp/Makefile
+++ b/lib/libxdp/Makefile
@@ -27,6 +27,11 @@ TEMPLATED_SOURCES := xdp-dispatcher.c
 CFLAGS += -I$(HEADER_DIR) -I$(LIB_DIR)/util
 BPF_CFLAGS += -I$(HEADER_DIR) -I$(STAGING_DIR)/usr/include
 
+ifeq ($(CONFIG_BIG_ENDIAN),y)
+	ENDIANNESS := "bpfeb"
+else
+	ENDIANESS := "bpfel"
+endif
 
 ifndef BUILD_STATIC_ONLY
 	SHARED_LIBS := $(OBJDIR)/libxdp.so \
@@ -127,7 +132,7 @@ $(TEMPLATED_SOURCES): %.c: %.c.in Makefi
 
 $(XDP_OBJS): %.o: %.c $(BPF_HEADERS) $(LIBMK)
 	$(QUIET_CLANG)$(CLANG) -S \
-	    -target bpf \
+	    -target $(ENDIANESS) \
 	    -D __BPF_TRACING__ \
 	    $(BPF_CFLAGS) \
 	    -Wall \
@@ -136,7 +141,7 @@ $(XDP_OBJS): %.o: %.c $(BPF_HEADERS) $(L
 	    -Wno-compare-distinct-pointer-types \
 	    -Werror \
 	    -O2 -emit-llvm -c -g -o ${@:.o=.ll} $<
-	$(QUIET_LLC)$(LLC) -march=bpf -filetype=obj -o $@ ${@:.o=.ll}
+	$(QUIET_LLC)$(LLC) -march=$(ENDIANESS) -filetype=obj -o $@ ${@:.o=.ll}
 
 .PHONY: man
 ifeq ($(EMACS),)
--- a/lib/common.mk
+++ b/lib/common.mk
@@ -20,6 +20,12 @@ XDP_OBJ_INSTALL ?= $(XDP_OBJ)
 LIB_DIR ?= ../lib
 LDLIBS ?= $(USER_LIBS)
 
+ifeq ($(CONFIG_BIG_ENDIAN),y)
+	ENDIANNESS := bpfeb
+else
+	ENDIANNESS := bpfel
+endif
+
 include $(LIB_DIR)/defines.mk
 include $(LIB_DIR)/libxdp/libxdp.mk
 
@@ -105,8 +111,8 @@ $(USER_TARGETS): %: %.c  $(OBJECT_LIBBPF
 	 $< $(LDLIBS)
 
 $(XDP_OBJ): %.o: %.c $(KERN_USER_H) $(EXTRA_DEPS) $(BPF_HEADERS) $(LIBMK)
-	$(QUIET_CLANG)$(CLANG) -S \
-	    -target bpf \
+	$(CLANG) -S \
+	    -target $(ENDIANNESS) \
 	    -D __BPF_TRACING__ \
 	    $(BPF_CFLAGS) \
 	    -Wall \
@@ -115,7 +121,7 @@ $(XDP_OBJ): %.o: %.c $(KERN_USER_H) $(EX
 	    -Wno-compare-distinct-pointer-types \
 	    -Werror \
 	    -O2 -emit-llvm -c -g -o ${@:.o=.ll} $<
-	$(QUIET_LLC)$(LLC) -march=bpf -filetype=obj -o $@ ${@:.o=.ll}
+	$(QUIET_LLC)$(LLC) -march=$(ENDIANNESS) -filetype=obj -o $@ ${@:.o=.ll}
 
 .PHONY: man
 ifeq ($(EMACS),)
